### PR TITLE
Attribute to retain cells during optimization

### PR DESCRIPTION
**Pull request by ayakayorihiro**
I added a new boolean attribute that indicates which cells (and uses of cells) should be preserved during optimizations. I check for the attribute in the `dead-cell-removal` pass, so the pass doesn't optimize those cells when it would otherwise do so. (@ekiwi and I got to the bottom of the errors I was experiencing yesterday)

The attribute is currently called `@protected`, but if anyone has suggestions about what the attribute should be named that would be helpful!

## Example
The test file `tests/passes/dead-cell-removal-protected.futil` differs from `tests/passes/dead-cell-removal.futil` by a single line, where we add `@protected` to unused register `unused_reg`:
```
$ diff tests/passes/dead-cell-removal.futil tests/passes/dead-cell-removal-protected.futil               
30c30                                                                                                                                                    
<     unused_reg = std_reg(32);                                                                                                                          
---                                                                                                                                                      
>     @protected unused_reg = std_reg(32);                                                                                                               
```
The expect files showing the result of running `dead-cell-removal` indicate that the pass retained `unused_reg` because of the `@protected` attribute:
```
$ diff tests/passes/dead-cell-removal.expect tests/passes/dead-cell-removal-protected.expect             
26a27                                                                                                                                                    
>     @protected unused_reg = std_reg(32); 
```

### CODE REVIEW

**Comment by rachitnigam** [comment]
Do you have thoughts on creating a more robust mechanism to ensure that `@protected` is respected when removing cells? Also, what should be the behavior when we're optimizing programs to share cells?

**Comment by ayakayorihiro** [comment]
Would including a check at the end of `dead-cell-removal` (and any other passes that can remove cells) that the `@protected` cells are preserved make sense for ensuring?
I think an approach to `cell-share` could be to simply say "cells cannot have both `protected` and `share`/`state_share` attributes" (and add a check at the beginning of `cell-share` for this)?

**Comment by ayakayorihiro** [comment]
After taking a closer look at the `cell-share` pass, I think `protected` should *override* `share`/`state_share`. So even though `std_wire` the primitive has the `share` attribute, cells of `std_wire` that have `protected` should *not* be shared. The `cell-share` pass currently has this behavior, and I added a test (`tests/passes/cell-share/protected.futil`) to check for this.

File: calyx-frontend/src/attribute.rs [hunk 1]
```diff
@@ -76,6 +76,9 @@ pub enum BoolAttr {
     #[strum(serialize = "fast")]
     /// https://github.com/calyxir/calyx/issues/1828
     Fast,
+    #[strum(serialize = "protected")]
+    /// Do I preserve this cell (and its uses) during optimization?
```
Comment (ekiwi) [inline calyx-frontend/src/attribute.rs:80]: Could you make this a little bit more clear? Maybe "Indicated that the cell should not be removed or shared during optimization".

**Comment by ekiwi** [comment]
This looks good now. We should merge this once we get an OK from @rachitnigam .

**Comment by rachitnigam** [comment]
Please add documentation for the new attribute (we should some day make this automatic): https://docs.calyxir.org/lang/attributes.html

Once that is done, go ahead with the merge!

**Pull request merged**